posts/2023-10-30_floating_point/floating_point.qmd

```diff
@@ -0,0 +1,338 @@
+
+As long as [{{admiral}}](https://github.com/pharmaverse/admiral) remains open source and free to use, using this package, or even reusing the code itself would be fine. 
+Although this was *my* preferred option, we did not end up implementing it.
+Instead, we made use of the `signif()` function, which rounds a number to a specified number of significant digits.
+This way, we could use the regular infix operators and simply provide the number of significant digits we want to compare to:
+
+```{r}
+signif_dig <- 15
+
+signif(AVAL, signif_dig) == signif(ANRHI * 1.1, signif_dig)
+
+# as:
+(ANRHI * 1.1) %>%
+  signif(signif_dig) %>%
+  format(digits = 22)
+
+# and although when printed, the number still looks off:
+ANRHI <- 101
+((ANRHI * 1.1) %>% signif(signif_dig)) %>% format(digits = 22)
+
+# the comparison works now:
+((ANRHI * 1.1) %>% signif(signif_dig)) == 111.1
+```
+
+This is now implemented throughout `atoxgr_criteria_ctcv5`:
+```{r}
+admiral::atoxgr_criteria_ctcv5 %>%
+  head() %>%
+  select(TERM, Definition, GRADE_CRITERIA_CODE)
+```
+
+## Conclusion
+<!-- The first version of this conclusion was generated with ChatGPT. -->
+
+In conclusion, the recent challenges faced by [{{admiral}}](https://github.com/pharmaverse/admiral) in dealing with floating point values shed light on the complexities and nuances of working with these numerical representations. Floating point values, as we've seen, are approximations of real numbers and can lead to unexpected issues in mathematical operations, especially when using exact comparators like `==` and `>=`. The differences between how these values are stored and computed can result in platform-specific discrepancies and unexpected behavior.
+
+Several potential solutions were explored to address this issue, including rounding, using `near()` or `all.equal()` functions, or implementing custom infix operators as seen in the fpCompare package. However, the most elegant and practical solution adopted in [{{admiral}}](https://github.com/pharmaverse/admiral) was to use the signif() function to round values to a specified number of significant digits. This approach allows for reliable and consistent comparisons without adding unnecessary complexity to the code base.
+
+Readers and developers should be vigilant when working with floating point values in their own code or when utilizing [{{admiral}}](https://github.com/pharmaverse/admiral) for their projects. Keep in mind that some floating point values can look like integers at first glance as in the above example of `1.1*100`. The experience with floating point issues in [{{admiral}}](https://github.com/pharmaverse/admiral) serves as a valuable reminder of the potential pitfalls associated with numerical precision in programming. It's crucial to exercise caution when performing comparisons with floating point numbers as small discrepancies can have significant downstream implications. When writing your own comparisons consider the following best practices:
```

Comment:
I would maybe make this a new section called "Best Practices"

Reply:
Good idea! 
I think I'll leave it as is, as the sections would get too small for my taste.